README.md
# Arctic Web Map

![samples/z5.jpg](samples/z5.jpg)

([More screenshots](samples))

These are the CartoCSS map stylesheets for the Canadian Arctic map layer on [airport.gswlab.ca][airport]. The style is based on [openstreetmap-carto][osm-carto] with some changes for the Canadian Arctic.

These stylesheets can be used in your own cartography projects, and are designed to be easily customised. They work with [Kosmtik][] and also with the command-line [CartoCSS][] processor.

If you want to use the map or tiles in your project, please see the [USAGE.md](USAGE.md) document. The rest of the README is for a general overview of the Arctic Web Map stylesheet and how to extend it.

[airport]: http://airport.gswlab.ca
[CartoCSS]: https://github.com/mapbox/carto
[Kosmtik]: https://github.com/kosmtik/kosmtik
[osm-carto]: https://github.com/gravitystorm/openstreetmap-carto

## Differences from Arctic Web Map [stable/1.0][stable-10]

The `stable/2.0` branch is a separate map style from the primary Arctic Web Map from ArcticConnect. Here is a comparison.

| Feature                       | `stable/1.0`           | `stable/2.0`                |
|-------------------------------|------------------------|-----------------------------|
| Active Development (2019)     | No                     | Yes                         |
| Coverage                      | Pan-Arctic             | Canada + Alaska + Greenland |
| Projections                   | EPSG:3571 to EPSG:3576 | EPSG:3573 only              |
| Bathymetry Styles             | Yes                    | Yes                         |
| Contour Lines                 | Canada only            | No                          |
| `openstreetmap-carto` Version | v2.45.1                | v4.20.0                     |
| Minutely OSM Updates          | Inactive               | No                          |

(Table generated with [Table Convert Online][table-convert]).

Currently, active development on the `master` branch will be merged into the `stable/2.0` branch.

[stable-10]: https://github.com/GeoSensorWebLab/awm-styles/tree/stable/1.0
[table-convert]: https://tableconvert.com

## Differences from openstreetmap-carto

* Web Mercator projection is replaced with a more appropriate projection for the far north (EPSG:3573)
* Bathymetry ocean data improves visualization of arctic region where oceans and seas significantly impact local communities and research
* Northern communities have increased label prominence (Cut-off is 60˚N)
* Region is clipped to -180˚E/40˚N to 0˚E/90˚N
* Hypsometric cross-tint is used as land base layer up to zoom 12
* Lakes from Natural Earth data are used up to zoom 7
* Map background is "land" colour with land/water polygons overlaid
* OSM `water-areas`, `text-poly-low-zoom`, `capital-names` not shown until zoom 5
* OSM `landcover-low-zoom`, `landcover-area-symbols` not shown until zoom 7
* OSM data in PostGIS is stored in EPSG:3573 (fixes label issues)
* OSM `country-names` removed to not over-emphasize *Kalaallit Nunaat*
* High-quality OSM split water polygons are used from zoom 10 instead of bathymetry

Other changes will be listed here as the stylesheet is developed.

# Installation

Installation has been separated into two documents, depending on your goal.

If you want to modify the stylesheet and commit changes to git, see [DEVELOPMENT.md](DEVELOPMENT.md).

For setting up the Arctic Web Map stylesheet on a public server, see [DEPLOYMENT.md](DEPLOYMENT.md).

# Map Errors

You may encounter issues with the map where data is not properly represented or is flat out incorrect. Some of these problems are with the stylesheet or data import process used by Arctic Web Map. Other problems are due to the underlying data from OpenStreetMap.

Problems with how things are drawn on the map probably fall under this stylesheet. Please use the issues tracker to notify us of any map problems.

Incorrect names, missing data, and similar are based in OpenStreetMap. You can fix small errors yourself with an OSM account, but for larger issues that affect large regions (city scale and larger) you should contact the OSM mailing list for confirmation.

# Contributing

Contributions to this project are welcome, see [CONTRIBUTING.md](CONTRIBUTING.md) for full details.

# Versioning

Arctic Web Map does not have an active versioning system. Branches are used to designate different variants of the stylesheet, as well as which map servers are on which version.

# Roadmap

The `stable/2.0` branch is intended to provide a general purpose stylesheet for researchers and northern communities. To that end, legibility of a wide variety of data from OpenStreetMap will be the focus.

# Alternatives

There are many open-source stylesheets written for creating OpenStreetMap-based maps using Mapnik, many based on the parent project. Some alternatives are:

* [OSM-Bright](https://github.com/mapbox/osm-bright)
* [XML-based stylesheets](https://trac.openstreetmap.org/browser/subversion/applications/rendering/mapnik)
* [osmfr-cartocss](https://github.com/cquest/osmfr-cartocss)
* [openstreetmap-carto-german](https://github.com/giggls/openstreetmap-carto-de)

# Maintainers

* James Badger [@openfirmware](https://github.com/openfirmware/)


CHANGELOG.md
# Release Notes

A list of changes made to the stylesheet.

### 2019-04-01

* Development branch for Canadian Arctic Web Map replaces `master` branch
* Uses `openstreetmap-carto` v4.20.0

### 2017-07-10

* Merge changes from v2.37.0
    * new farmland color - makes farmland less prominent
    * Adding shop=copyshop icon
    * Drop incorrect text class from text-line
    * Change order of definitions to speed up carto processing
    * Restore rendering of water-feature labels
    * Make residential and tertiary slightly narrower on z13
    * Osmic icon for shop=beverages
    * correct variable name for path/footway tunnel handling
    * Make roads narrower at zoom 18 and 19
    * rework admin boundaries and admin labels on lower zooms, render country subdivisions only for the biggest * s on z4
    * Add SQL conditions for low level admin lines and labels
    * Don't use centroid for low-zoom admin calcs
    * remove admin\_level 0,1,3 from label rendering and consolidate table aliases
* Merge changes from v2.38.0
    * Code change for man\_made=cutline
    * Increase size of small place names on z15 by one pixel
    * Icon and YAML for amenity=social\_facility
    * Revert "rework admin boundaries and admin labels on lower zooms"
    * rework admin boundaries (without z4 logic)
    * Adding rendering of shoal and reef Rendering surface on shoal and beach through pattern variation
    * updating surface values based on actual use
    * remove line barrier catch-all rendering
    * Adding code and icon for shop=sports
* Merge changes from v2.39.0
    * Replacing shelter PNG icon with SVG version
    * Replacing cave entry PNGicon  with SVG version
    * fix oneway arrow colour for highway=path/footway and bicycle=designated/horse=designated
    * render swing\_gate like lift\_gate
    * Drop 'disused' from stations SQL
    * Replacing tourism information PNG icon with SVG version
* Merge changes from v2.40.0
    * Reformat road SQL
    * Add water polygons to the files to be downloaded
    * Replace land shapefiles with ocean shapefiles in rendering
    * Remove land polygon files
    * Move water layers to their old position
    * Refactor water area MSS
    * Make tertiary road yellowish
    * explain purpose of scripts in /scripts
    * attempt to clarify purpose of shop\_values.rb fixes part of #2080
    * Revert "Make tertiary road yellowish"
    * Remove outdated references to land polygons
    * Add custom index information
    * Mention that some issues are tagged as easy
    * Fix bug that prevents inaccessible tracks on bridges from being rendered
    * Remove IF NOT EXISTS for indexes
    * Make place ordering better defined
    * Pull in upstream default.style changes
    * Use a script to create indexes, with adjustable options
    * Remove unnecessary index and duplicate WHERE
    * Add a way\_area index for z6 performance
    * Add guideline on symbolizers which change per zoom.
    * exclude shop values that should no be rendered
    * drop unneeded whitespace
    * drop dead shop=fishmonger code
    * wider rails on z18, wider service rail at z18+
    * introduce @standard-text-size, resize 11, 9, 8 sized POI text labels to 10
    * Replacing library PNG icon with SVG version
    * fix positioning of icon
    * set text dy to 11
    * render shop=fishmonger, fixes #2155
    * Add label for tourism=viewpoint areas
    * more readable code, the same selector style for all turning circles
    * fix  missing forest pattern on forests with unexpected natural key
    * shop=books reusing library icon with the shop color
    * thinner power=line and power=minor\_line
    * darker standard rail on higher zoom
    * resize turning circles automagically, make customization easier, fixes #1931
    * placenames-medium is not used on z3
    * Revert "Merge remote-tracking branch 'pnorman/water\_polygons'"
    * rewrite low priority layer to make it like main amenity layer
    * stop displaying highway=mini\_roundabout and highway=level\_crossing mapped as areas
    * larger labels for placenames
    * Filter placenames-large based on way\_area
    * Add name IS NOT NULL check to medium placenames
    * synchronise description in Github repo and project file
    * it is not necessary to query buildings for z12, as buildings are not displayed on z12
    * show bus\_guideway tunnels
* Merge changes from v2.41.0
    * changing sports related colors
    * changing color name from @educational\_areas\_and\_hospital to @societal\_amenities
    * fix possible admin label and line rendering mismatch by changing admin label table
    * show ref for aeroway=gate and ele for peaks/volcano/saddle/alpine\_hut without names
    * remove text-min-distance - increase label density
    * switch to book font for POI labels
    * add @standard-halo-radius
    * Remove Mallige from font list
    * Replace ttf-indic-fonts-core with new package names
    * Use new font package for DejaVu Sans
    * Remove font instructions for Ubuntu 13.10
    * Reformat font install line
    * Font documentation wording improvements
    * Add Tibetan Machine Uni to documentation, as it was missing
    * Remove fonts not present in installed packages
    * Update droid font info to Ubuntu 16.04
    * Add new name for Arundina fonts
    * Add notes about Debian Testing font packages
    * render man\_made=obelisk, fixes #2126
    * Add rendering of leisure=dog\_park
    * Split up roads.mss, refactor road colour script
    * Use argparse for arguments
    * Move all road colour printing together
    * Use 2sp in road-colors.yaml to be the same as other yaml
    * Move colour name generation later
    * Let generate\_shields use road-colors.yaml
    * Use binary mode for mml in yaml2mml
* Merge changes from v2.42.0
    * Add information on how to make a release
    * Update releasing with information on schema change and reload
    * add amenity=charging\_station icon
    * SVG version of department store icon

awmconfig.js
const fs  = require("fs");
const pry = require("pryjs");

const epsg4326 = "+proj=longlat +datum=WGS84 +no_defs";
const epsg3573 = "+proj=laea +lat_0=90 +lon_0=-100 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs";
// maximum extent of EPSG:3573 projection
const extent3573 = [
    -20037508,
    -20037508,
    20037508,
    20037508
];
// clipped extent of EPSG:3573 projection for this map
const clipped3573 = [
    -4167630.8211728190071881,
    -5397732.1324295047670603,
    5396780.7497062487527728,
    937306.5764416041783988
];

// TODO: Create a DSL/API to clean this up.
// AddAfter, Remove, CreateLayer, etc.
exports.LocalConfig = function (localizer, project) {
    /*
     * PROJECT RECONFIGURATION
     */
    project.mml.name = "ArcticWebMap";
    project.mml.center = [0, 0, 3];
    project.mml.bounds = extent3573;
    // Use EPSG:3573 PROJ4 string
    project.mml.srs = epsg3573;

    // extents from ne_10m_land projected into EPSG:3573
    project.mml["maximum-extent"] = clipped3573;

    // Cache features in RAM. Uses a lot of RAM that would better be
    // used by Postgres.
    // localizer.where("Layer")
    // .then({
    //     "cache-features": "on"
    // });

    /*
     * LAYER REMOVALS
     */

    // remove antarctic layers - projection issues
    // remove coasts - high-zoom coastline not needed
    // remove country names - only Kalaallit Nunaat/France/Île Verte are
    // displayed, which may look like favouritism to those regions
    function remove(searchID) {
        let index = project.mml.Layer.findIndex((l) => (l.id === searchID));
        project.mml.Layer.splice(index, 1);
    };

    removedIDs = [
        "coast-poly",
        "country-names",
        "icesheet-outlines",
        "icesheet-poly"
    ].forEach((id) => { remove(id) });

    /*
     * LAYER EDITS
     */

    // Change database name, details, projection
    localizer.where("Layer")
    .if({"Datasource.type": "postgis"})
    .then({
        "Datasource.dbname": "osm_3573",
        "Datasource.password": "",
        "Datasource.user": "",
        "Datasource.host": "",
        "Datasource.persist_connection": true,
        "Datasource.estimate_extent": true,
        // extent should be in database SRS
        "Datasource.extent": clipped3573.join(", "),
        "srs-name": "EPSG:3573",
        "srs": epsg3573
    });

    // Use natural earth 10m for land instead of OSM
    localizer.where("Layer")
    .if({"Datasource.file": "data/simplified-land-polygons-complete-3857/simplified_land_polygons.shp"})
    .then({
        "Datasource.file": "data/ne_10m_land/ne_10m_land.shp"
    });

    /*
     * MINZOOM ADJUSTMENTS
     */

    // Adjust minzoom on water-areas
    localizer.where("Layer")
    .if({"id": "water-areas"})
    .then({
        "properties.minzoom": 5
    });

    // Adjust minzoom on landcover-low-zoom
    localizer.where("Layer")
    .if({"id": "landcover-low-zoom"})
    .then({
        "properties.minzoom": 7
    });

    // Adjust minzoom on landcover-area-symbols
    localizer.where("Layer")
    .if({"id": "landcover-area-symbols"})
    .then({
        "properties.minzoom": 7
    });

    // Adjust minzoom on text-poly-low-zoom
    localizer.where("Layer")
    .if({"id": "text-poly-low-zoom"})
    .then({
        "properties.minzoom": 5
    });

    // Adjust minzoom on capital-names
    localizer.where("Layer")
    .if({"id": "capital-names"})
    .then({
        "properties.minzoom": 5
    });

    // Adjust bias for cities, towns, and villages north of 60˚N.
    // For places in that region:
    // 1. classify in "city" category
    // 2. multiply population by 100
    // 3. Include villages in this layer
    localizer.where("Layer")
    .if({"id": "placenames-medium"})
    .then({"Datasource.table":
    `(SELECT
            way,
            name,
            score,
            CASE
              WHEN (place = 'city') THEN 1
              WHEN (ST_Covers(ST_MakeEnvelope(-180,60,0,90,4326), ST_Transform(way, 4326))) THEN 1
              WHEN (place = 'town') THEN 2
              ELSE 3
            END as category,
            round(ascii(md5(osm_id::text)) / 55) AS dir -- base direction factor on geometry to be consistent across metatiles
          FROM
            (SELECT
                osm_id,
                way,
                place,
                name,
                (
                  (CASE
                    WHEN (tags->'population' ~ '^[0-9]{1,8}$') THEN (tags->'population')::INTEGER
                    WHEN (place = 'city') THEN 100000
                    WHEN (place = 'town') THEN 1000
                    WHEN (place = 'village') THEN 100
                    ELSE 1
                  END)
                  *
                  (CASE
                    WHEN (ST_Covers(ST_MakeEnvelope(-180,60,0,90,4326), ST_Transform(way, 4326))) THEN 100
                    WHEN (tags @> 'capital=>4') THEN 2
                    ELSE 1
                  END)
                ) AS score
              FROM planet_osm_point
              WHERE place IN ('city', 'town', 'village')
                AND name IS NOT NULL
                AND NOT (tags @> 'capital=>yes')
            ) as p
          ORDER BY score DESC, length(name) DESC, name
        ) AS placenames_medium`,
        "properties.minzoom": 5});

    // Adjust shapefile layer file location, source projection
    localizer.where("Layer")
    .if({"Datasource.type": "shape"})
    .then((obj) => {
        obj["srs-name"] = "EPSG:3573";
        obj["srs"] = epsg3573;
        obj.extent = clipped3573;
        obj.Datasource.extent = clipped3573;
        obj.Datasource.file = obj.Datasource.file.replace(/^data/, "data/awm");
        return obj;
    });

    /*
     * LAYER ADDITIONS
     */

    // Add a layer to the Layer list after the layer with id `searchID`
    function addAfter(searchID, layer) {
        let index = project.mml.Layer.findIndex(l => (l.id === searchID));
        project.mml.Layer.splice(index+1, 0, layer);
    }

    // Sample layer format:
    // {
    //     "id": "world",
    //     "geometry": "polygon",
    //     "extent": [
    //         -20037508,
    //         -20037508,
    //         20037508,
    //         20037508
    //     ],
    //     "srs-name": "EPSG:3573",
    //     "srs": "+proj=laea +lat_0=90 +lon_0=-100 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs",
    //     "Datasource": {
    //         "file": "data/awm/ne_10m_land/ne_10m_land.shp",
    //         "type": "shape",
    //         "extent": [
    //             -20037508,
    //             -20037508,
    //             20037508,
    //             20037508
    //         ]
    //     },
    //     "properties": {
    //         "maxzoom": 9
    //     }
    // }
    
    addAfter("world", {
        id: "world-tint",
        geometry: "raster",
        "srs-name": "EPSG:3573",
        srs: epsg3573,
        Datasource: {
            file: "data/awm/HYP_HR_SR/HYP_HR_SR.tif",
            type: "gdal",
            band: -1
        },
        properties: {
            maxzoom: 12
        }
    });

    addAfter("necountries", {
        id: "lakes-low",
        geometry: "polygon",
        "srs-name": "EPSG:3573",
        srs: epsg3573,
        Datasource: {
            file: "data/awm/ne_10m_lakes/ne_10m_lakes.shp",
            type: "shape"
        },
        properties: {
            maxzoom: 7
        }
    });

    // Add bathymetry layers
    let bathymetry = {
        "bathymetry-10km": "ne_10m_bathymetry_A_10000",
        "bathymetry-9km":  "ne_10m_bathymetry_B_9000",
        "bathymetry-8km":  "ne_10m_bathymetry_C_8000",
        "bathymetry-7km":  "ne_10m_bathymetry_D_7000",
        "bathymetry-6km":  "ne_10m_bathymetry_E_6000",
        "bathymetry-5km":  "ne_10m_bathymetry_F_5000",
        "bathymetry-4km":  "ne_10m_bathymetry_G_4000",
        "bathymetry-3km":  "ne_10m_bathymetry_H_3000",
        "bathymetry-2km":  "ne_10m_bathymetry_I_2000",
        "bathymetry-1km":  "ne_10m_bathymetry_J_1000",
        "bathymetry-200m": "ne_10m_bathymetry_K_200",
        "bathymetry-0m":   "ne_10m_bathymetry_L_0"
    }

    Object.keys(bathymetry).forEach((id) => {
        let filename = bathymetry[id];
        addAfter("necountries", {
            id: id,
            geometry: "polygon",
            "srs-name": "EPSG:3573",
            srs: epsg3573,
            Datasource: {
                file: `data/awm/ne_10m_bathymetry_all/${filename}.shp`,
                type: "shape"
            },
            properties: {
                maxzoom: 9
            }
        });
    });

    // Add high-quality water layer for mid and high zoom levels
    addAfter("bathymetry-0m", {
        id: "water-poly",
        geometry: "polygon",
        "srs-name": "EPSG:4326",
        srs: epsg4326,
        Datasource: {
            file: "data/water_polygons/water-polygons-split-4326/water_polygons.shp",
            type: "shape"
        },
        properties: {
            minzoom: 10
        }
    });

    /*
     * STYLESHEET ADDITIONS
     */

    // Add to stylesheets
    project.mml.Stylesheet.push({
        id: "../arcticwebmap.mss",
        data: fs.readFileSync("./arcticwebmap.mss").toString()
    });
};
